Source/WebKit/UIProcess/WebPageProxy.h

```diff
@@ -2917,10 +2918,39 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool m_isListeningForUserFacingStateChangeNotification { false };
 #endif
     bool m_allowsMediaDocumentInlinePlayback { false };
-    std::unique_ptr<ProcessThrottlerActivity> m_isVisibleActivity;
-    std::unique_ptr<ProcessThrottlerActivity> m_isAudibleActivity;
-    std::unique_ptr<ProcessThrottlerActivity> m_isCapturingActivity;
-    std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
+
+    struct ProcessActivityState {
+        WTF_MAKE_STRUCT_FAST_ALLOCATED;
+
+        WebPageProxy& m_page;
+
+        std::unique_ptr<ProcessThrottlerActivity> m_isVisibleActivity;
+        UniqueRef<ProcessThrottlerTimedActivity> m_wasRecentlyVisibleActivity;
+        std::unique_ptr<ProcessThrottlerActivity> m_isAudibleActivity;
+        std::unique_ptr<ProcessThrottlerActivity> m_isCapturingActivity;
+        std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
+
+        ProcessActivityState(WebPageProxy&);
+        void takeVisibleActivity();
+        void takeAudibleActivity();
+        void takeCapturingActivity();
+        void takeOpeningAppLinkActivity();
+
+        void reset();
+        void dropVisibleActivity();
+        void dropAudibleActivity();
+        void dropCapturingActivity();
+        void dropOpeningAppLinkActivity();
+
+        bool hasValidVisibleActivity() const;
+        bool hasValidAudibleActivity() const;
+        bool hasValidCapturingActivity() const;
+        bool hasValidOpeningAppLinkActivity() const;
+    };
+
+    UniqueRef<ProcessActivityState> m_processActivityState;
+    ProcessActivityState& processActivityState() { return m_processActivityState; }
```

Comment:
Getters should be in the middle of data members. Also not sure these getters are terribly useful. We could access m_processActivityState directly.